File: backend/src/main/java/com/flowlink/execution/engine/TcpNodeExecutor.java
package com.flowlink.execution.engine;

import com.flowlink.common.json.JsonService;
import com.flowlink.core.graph.GraphNode;
import com.flowlink.core.graph.TcpField;
import com.flowlink.core.graph.TcpRespField;
import org.springframework.stereotype.Component;

import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * 고정길이 금융 전문 TCP 노드 실행기.
 * 요청 필드를 바이트 단위 고정길이로 조립해 길이-프리픽스 전문으로 전송하고,
 * 응답 전문을 응답 필드 길이대로 잘라 출력으로 만든다. (인코딩 노드/필드별 선택)
 */
@Component
public class TcpNodeExecutor {

    private final TokenResolver tokens;
    private final SsrfGuard ssrfGuard;
    private final JsonService json;

    public TcpNodeExecutor(TokenResolver tokens, SsrfGuard ssrfGuard, JsonService json) {
        this.tokens = tokens;
        this.ssrfGuard = ssrfGuard;
        this.json = json;
    }

    public NodeResult execute(GraphNode node, ExecutionContext ctx) {
        String host = tokens.resolveTokens(node.tcpHost() == null ? "" : node.tcpHost(), ctx);
        int port = node.tcpPort() == null ? 0 : node.tcpPort();
        Charset nodeCs = charset(node.tcpEncoding(), Charset.forName("EUC-KR"));
        int timeout = node.tcpTimeoutMs() == null || node.tcpTimeoutMs() <= 0 ? 5000 : node.tcpTimeoutMs();
        int prefixLen = node.tcpPrefixLength() == null ? 0 : node.tcpPrefixLength();
        boolean includesSelf = Boolean.TRUE.equals(node.tcpPrefixIncludesSelf());

        // 1) 요청 전문 본문 조립(바이트 고정길이)
        Map<String, Object> reqValues = new LinkedHashMap<>();
        java.io.ByteArrayOutputStream bodyBuf = new java.io.ByteArrayOutputStream();
        for (TcpField f : node.tcpRequest() == null ? List.<TcpField>of() : node.tcpRequest()) {
            String v = resolveField(f, ctx);
            if (f.name() != null && !f.name().isBlank()) {
                reqValues.put(f.name(), v);
            }
            Charset cs = charset(f.encoding(), nodeCs);
            byte[] field = fixedField(v, f.lengthOrZero(), f.pad(), f.padChar(), cs);
            bodyBuf.writeBytes(field);
        }
        byte[] body = bodyBuf.toByteArray();

        // 2) 길이 프리픽스
        byte[] message;
        if (prefixLen > 0) {
            int declared = includesSelf ? body.length + prefixLen : body.length;
            byte[] prefix = prefix(declared, prefixLen);
            byte[] m = new byte[prefix.length + body.length];
            System.arraycopy(prefix, 0, m, 0, prefix.length);
            System.arraycopy(body, 0, m, prefix.length, body.length);
            message = m;
        } else {
            message = body;
        }

        String reqText = "TCP " + host + ":" + port + " (" + nodeCs.name() + ", " + message.length + "B)\n" + printable(message, nodeCs);

        // 3) SSRF
        try {
            ssrfGuard.checkHostPort(host, port);
        } catch (SsrfBlockedException e) {
            return NodeResult.fail(0, reqText, "⚠ 차단됨(SSRF 가드): " + e.getMessage());
        }

        // 4) 송수신
        try (Socket socket = new Socket()) {
            socket.connect(new InetSocketAddress(host, port), timeout);
            socket.setSoTimeout(timeout);
            OutputStream out = socket.getOutputStream();
            out.write(message);
            out.flush();

            InputStream in = socket.getInputStream();
            byte[] respBody;
            if (prefixLen > 0) {
                byte[] pre = readN(in, prefixLen);
                int declared;
                try {
                    declared = Integer.parseInt(new String(pre, StandardCharsets.US_ASCII).trim());
                } catch (NumberFormatException e) {
                    return NodeResult.fail(0, reqText, "⚠ 응답 길이 프리픽스 파싱 실패: '" + new String(pre, StandardCharsets.US_ASCII) + "'");
                }
                int bodyLen = includesSelf ? declared - prefixLen : declared;
                if (bodyLen < 0) {
                    return NodeResult.fail(0, reqText, "⚠ 잘못된 응답 길이: " + declared);
                }
                respBody = readN(in, bodyLen);
            } else {
                respBody = in.readAllBytes();
            }

            // 5) 응답 슬라이싱 → 출력
            Map<String, Object> value = new LinkedHashMap<>();
            int offset = 0;
            for (TcpRespField rf : node.tcpResponse() == null ? List.<TcpRespField>of() : node.tcpResponse()) {
                int len = rf.lengthOrZero();
                int end = Math.min(offset + len, respBody.length);
                byte[] slice = Arrays.copyOfRange(respBody, Math.min(offset, respBody.length), end);
                String decoded = new String(slice, charset(rf.encoding(), nodeCs));
                if (rf.name() != null && !rf.name().isBlank()) {
                    value.put(rf.name(), decoded);
                }
                offset += len;
            }
            String resText = "응답 " + respBody.length + "B\n" + printable(respBody, nodeCs);
            return new NodeResult(true, null, reqText, resText, value, value, reqValues, null);
        } catch (Exception e) {
            return NodeResult.fail(0, reqText, "⚠ TCP 요청 실패: " + (e.getMessage() == null ? e.toString() : e.getMessage()));
        }
    }

    private String resolveField(TcpField f, ExecutionContext ctx) {
        if (f.bound() != null) {
            return tokens.stringify(tokens.resolveBinding(f.bound(), ctx));
        }
        String v = f.value();
        return (v != null && v.contains("{{")) ? tokens.resolveTokens(v, ctx) : (v == null ? "" : v);
    }

    private static byte[] readN(InputStream in, int n) throws java.io.IOException {
        byte[] buf = in.readNBytes(n);
        if (buf.length < n) {
            throw new java.io.IOException("응답이 조기 종료됨 (" + buf.length + "/" + n + " 바이트)");
        }
        return buf;
    }

    private static byte[] prefix(int declared, int width) {
        String s = String.format("%0" + width + "d", declared);
        byte[] b = s.getBytes(StandardCharsets.US_ASCII);
        if (b.length > width) {
            // 길이 오버플로 — 하위 width 자리만(방어적)
            return Arrays.copyOfRange(b, b.length - width, b.length);
        }
        return b;
    }

    private static byte[] fixedField(String value, int length, String pad, String padChar, Charset cs) {
        byte[] raw = (value == null ? "" : value).getBytes(cs);
        if (length <= 0) {
            return new byte[0];
        }
        if (raw.length == length) {
            return raw;
        }
        if (raw.length > length) {
            return Arrays.copyOf(raw, length); // 초과 시 절단
        }
        byte[] out = new byte[length];
        byte pb = padByte(padChar, cs);
        if ("left".equalsIgnoreCase(pad)) {
            int padCount = length - raw.length;
            Arrays.fill(out, 0, padCount, pb);
            System.arraycopy(raw, 0, out, padCount, raw.length);
        } else { // 기본 right
            System.arraycopy(raw, 0, out, 0, raw.length);
            Arrays.fill(out, raw.length, length, pb);
        }
        return out;
    }

    private static byte padByte(String padChar, Charset cs) {
        String p = (padChar == null || padChar.isEmpty()) ? " " : padChar;
        byte[] b = p.getBytes(cs);
        return b.length > 0 ? b[0] : (byte) ' ';
    }

    private static Charset charset(String name, Charset def) {
        if (name == null || name.isBlank()) {
            return def;
        }
        try {
            return Charset.forName(name);
        } catch (Exception e) {
            return def;
        }
    }

    private static String printable(byte[] bytes, Charset cs) {
        String s = new String(bytes, cs);
        StringBuilder sb = new StringBuilder(s.length());
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            sb.append(c < 0x20 ? '.' : c);
        }
        return sb.toString();
    }
}

File: backend/src/main/kotlin/com/flowlink/execution/engine/TcpNodeExecutor.kt
package com.flowlink.execution.engine

import com.flowlink.common.error.NotFoundException
import com.flowlink.common.tcp.TcpBytes
import com.flowlink.core.graph.GraphNode
import com.flowlink.protocol.Direction
import com.flowlink.protocol.Framer
import com.flowlink.protocol.ProtocolCodec
import com.flowlink.protocol.ProtocolDtos
import com.flowlink.protocol.ProtocolService
import com.flowlink.protocol.ProtocolSpec
import com.flowlink.protocol.TcpClient
import org.springframework.stereotype.Component
import java.util.UUID

/**
 * TCP 요청 노드 — "프로토콜에 값을 채운 것". 조립/해석은 ProtocolCodec, 왕복은 TcpClient.
 * 출력 = 응답 헤더 + 본문 필드(정의되지 않은 응답 전문이면 헤더 + `body` raw 텍스트, 노드는 성공).
 */
@Component
class TcpNodeExecutor(private val tokens: TokenResolver, private val protocols: ProtocolService) {

    class Built(val spec: ProtocolSpec, val key: String, val encoded: ProtocolCodec.Encoded, val values: LinkedHashMap<String, String>, val host: String, val port: Int, val timeoutMs: Int, val reqText: String)

    fun build(node: GraphNode, ctx: ExecutionContext): Built {
        val pid = node.protocolId?.trim()?.takeIf { it.isNotEmpty() } ?: throw IllegalArgumentException("프로토콜을 선택하세요.")
        val spec = try { protocols.specOf(UUID.fromString(pid)) } catch (e: Exception) {
            if (e is NotFoundException || e is IllegalArgumentException) throw IllegalArgumentException("프로토콜을 찾을 수 없습니다: $pid") else throw e
        }
        val key = node.tcpMessage?.trim()?.takeIf { it.isNotEmpty() } ?: throw IllegalArgumentException("전문(거래코드)을 선택하세요.")
        val host = tokens.resolveTokens(node.tcpHost ?: "", ctx).trim()
        val port = node.tcpPort ?: 0
        val values = LinkedHashMap<String, String>()
        for ((k, v) in node.tcpValues ?: emptyMap()) values[k] = if (v.contains("{{")) tokens.stringify(tokens.resolveLiteral(v, ctx)) else v
        val enc = try {
            ProtocolCodec.encode(spec, key, values, Direction.SEND, protocols.plugins())
        } catch (e: ProtocolCodec.ProtocolException) {
            throw IllegalArgumentException(e.message, e)
        } catch (e: RuntimeException) {
            throw IllegalArgumentException("전문 조립 중 오류: ${e.message ?: e}", e)
        }
        val reqText = "TCP $host:$port · $key · ${spec.charset().name()} · ${enc.bytes.size}B\n" + table(enc.fields) + "\n" + TcpBytes.printable(enc.bytes, spec.charset())
        return Built(spec, key, enc, values, host, port, if ((node.tcpTimeoutMs ?: 0) <= 0) 5000 else node.tcpTimeoutMs!!, reqText)
    }

    /** 전송 없이 조립 — 검증/조립 실패는 errors 로(속성 패널이 필드 옆에 표시). */
    fun preview(node: GraphNode, ctx: ExecutionContext): ProtocolDtos.PreviewResult = try {
        val b = build(node, ctx)
        ProtocolDtos.PreviewResult(b.encoded.bytes.size, TcpBytes.hexDump(b.encoded.bytes), TcpBytes.printable(b.encoded.bytes, b.spec.charset()), b.encoded.fields, emptyList(), b.encoded.warnings)
    } catch (e: Exception) {
        val pe = e.cause as? ProtocolCodec.ProtocolException
        ProtocolDtos.PreviewResult(0, "", "", emptyList(), listOf(ProtocolDtos.PreviewError(pe?.field, e.message ?: "조립 실패")))
    }

    fun execute(node: GraphNode, ctx: ExecutionContext): NodeResult {
        val b = try { build(node, ctx) } catch (e: Exception) {
            return NodeResult.fail(0, "TCP ${node.tcpHost ?: ""}:${node.tcpPort ?: 0} · ${node.tcpMessage ?: ""}", "⚠ TCP 전문 조립 실패: " + (e.message ?: e.toString()))
        }
        if (b.host.isBlank()) return NodeResult.fail(0, b.reqText, "⚠ 호스트가 없습니다.")
        return try {
            val x = TcpClient.exchange(b.host, b.port, b.timeoutMs, b.encoded.bytes, b.spec)
            val d = ProtocolCodec.decode(b.spec, x.response.bytes, Direction.RECV, protocols.plugins())
            val out = LinkedHashMap<String, Any?>()
            out.putAll(d.header)
            if (d.body != null) out.putAll(d.body) else out["body"] = TcpBytes.decodeEscaped(d.rawBody, b.spec.charset())
            val sb = StringBuilder("응답 ${x.response.bytes.size}B · ${x.elapsedMs}ms")
            if (d.messageKey == null) sb.append(" · 정의되지 않은 전문 '${d.disc ?: ""}' — 본문 raw(body)")
            if (x.response.chunks.size > 1) sb.append(" · 부분 수신 ${x.response.chunks.joinToString("+")}B")
            for (w in d.warnings) sb.append("\n⚠ ").append(w)
            sb.append('\n').append(table(d.fields)).append('\n').append(TcpBytes.printable(x.response.bytes, b.spec.charset()))
            NodeResult(true, null, b.reqText, sb.toString(), out, out, LinkedHashMap<String, Any?>(b.values), null)
        } catch (e: Framer.FrameException) {
            NodeResult.fail(0, b.reqText, "⚠ 응답 프레이밍 실패: ${e.message}" + (e.raw?.let { "\nHEX " + TcpBytes.hexDump(it) } ?: ""))
        } catch (e: Exception) {
            NodeResult.fail(0, b.reqText, "⚠ TCP 요청 실패: " + (e.message ?: e.toString()))
        }
    }

    private fun table(fields: List<ProtocolCodec.FieldSlice>): String =
        fields.joinToString("\n") { "@${it.offset.toString().padStart(4)} ${it.name} = ${it.value}" + (it.warn?.let { w -> "  $w" } ?: "") }
}
